sodetlib/operations/bias_wave.py

```diff
@@ -0,0 +1,326 @@
+
+    if dc_bias is None:
+        dc_bias = S.get_tes_bias_bipolar_array()[bias_group]
+
+    for freq in freqs_wave:
+        S.log(f"BL sine wave with bg={bias_group}, freq={freq}")
+        S.play_sine_tes(bias_group, amp_wave, freq, dc_amp=dc_bias)
+        start_times.append(time.time())
+        time.sleep(duration)
+        stop_times.append(time.time())
+        S.set_rtm_arb_waveform_enable(0)
+        S.set_tes_bias_bipolar(bias_group, dc_bias)
+    return start_times, stop_times
+
+class BiasWaveAnalysis:
+    """
+    UPDATE THE DOCSTRING
+    """
+    def __init__(self, S=None, cfg=None, bgs=None, run_kwargs=None):
+        self._S = S
+        self._cfg = cfg
+
+        self.bgs = bgs
+        self.am = None
+        # WHAT DO I REPLACE THIS WITH??
+        # self.edge_idxs = None
+        
+        # DOES BEN USE THIS?
+        # self.transition_range = None
+
+        if S is not None:
+            self.meta = sdl.get_metadata(S, cfg)
+            self.stream_id = cfg.stream_id
+
+            if run_kwargs is None:
+                run_kwargs = {}
+            self.run_kwargs = run_kwargs
+            self.high_current_mode = run_kwargs.get("high_current_mode", True)
+
+    def save(self, path=None):
+        data = {}
+        saved_fields = [
+            # Run data and metadata
+            'bands', 'channels', 'sid', 'meta', 'run_kwargs', 'start', 'stop',
+            'high_current_mode', 'start_times', 'stop_times',
+            # Bgmap data
+            'bgmap', 'polarity',
+            # Step data and fits
+            'resp_times', 'mean_resp', 'step_resp',
+            # Step fit data
+            'step_fit_tmin', 'step_fit_popts', 'step_fit_pcovs',
+            'tau_eff',
+            # Det param data
+            'transition_range', 'Ibias', 'Vbias', 'dIbias', 'dVbias', 'dItes',
+            'R0', 'I0', 'Pj', 'Si',
+            # From IV's
+            'R_n_IV', 'Rfrac',
+        ]
+
+        for f in saved_fields:
+            if not hasattr(self, f):
+                print(f"WARNING: field {f} does not exist... "
+                      "defaulting to None")
+            data[f] = getattr(self, f, None)
+
+        if path is not None:
+            np.save(path, data, allow_pickle=True)
+            self.filepath = path
+        else:
+            self.filepath = sdl.validate_and_save(
+                'bias_wave_analysis.npy', data, S=self._S, cfg=self._cfg,
+                make_path=True
+            )
+
+    @classmethod
+    def load(cls, filepath):
+        self = cls()
+        data = np.load(filepath, allow_pickle=True).item()
+        for k, v in data.items():
+            setattr(self, k, v)
+        self.filepath = filepath
+        return self
+    
+    def run_analysis(self, assignment_thresh=0.3, arc=None, base_dir='/data/so/timestreams', 
+                     step_window=0.03, fit_tmin=1.5e-3, transition=None, R0_thresh=30e-3,
+                     save=False, bg_map_file=None):
```

Comment:
so one of the things worth checking is that when Yuhan and I wrote up our analysis notebooks we just used the bgmap from the most recent biasstep, usually was taken from before hwp spin-up. just wanted to confirm that is  what this code is able to do, or did you write it to generate a bgmap from the biaswave?

Reply:
See lines 109-114, if the argument `bg_map_file` provides a path to a bgmap then it will use that if not it will load the bgmap which is in the device config file which should be the most recent bgmap (which if you setup your night of observing this is likely also before you spun up the hwp for the night). There's no attempt to calculate a bgmap from the bias wave measurements here.